### PR TITLE
Yang/issue#1

### PR DESCRIPTION
**Pull request by thomasly**
The model is pretrainable now.

Major changes:
1. in dataloader.py, added a DataLoaderPubchemContext to handle the new dataset format with substructs. The new data loader uses a new collate function, BatchPubchemContext (in batch.py), to add some attributes to the mini-batch.
2. in dataloader.py, added an ExtractPubchemSubstructs, which is majorly the same as the original ExtractSubstructureContextPair class. The x_substruct is based on a root_substructure instead of a root_atom now.
3. the new training logic is in pretrain_contextsub.py. The way to handle outputs from the context model and the substructure model is slightly changed.
4. the MoleculeDataset is completely refactored. The main logic remains the same. Each Data object now contains a substructs attribute.
5. move all helper functions in loader.py to util.py. Move data loading related functions/classes from util.py to dataloader.py.

Minor changes include but not limited to linting and a few changes to the variable names.

### CODE REVIEW

**Comment by thomasly** [comment]
Jimmy,
If you have questions, we can have a discussion whenever you have time. Please write unit tests to make sure my new data are correctly generated and handled during the training process if possible. Thank you!

**Comment by liketheflower** [comment]
Thanks. I am reviewing this code now.

File: batch.py [hunk 1]
```diff
@@ -142,19 +149,28 @@ def from_data_list(data_list):
         r"""Constructs a batch object from a python list holding
         :class:`torch_geometric.data.Data` objects.
         The assignment vector :obj:`batch` is created on the fly."""
-        #keys = [set(data.keys) for data in data_list]
-        #keys = list(set.union(*keys))
-        #assert 'batch' not in keys
+        # keys = [set(data.keys) for data in data_list]
```
Comment (liketheflower) [inline batch.py:152]: line 152 to 154  can be deleted.

File: batch.py [hunk 2]
```diff
@@ -142,19 +149,28 @@ def from_data_list(data_list):
...
+            "x_substruct",
+            "overlap_context_substruct_idx",
+            "edge_attr_context",
+            "edge_index_context",
+            "x_context",
+        ]
 
         for key in keys:
-            #print(key)
+            # print(key)
```
Comment (liketheflower) [inline batch.py:169]: This line can be deleted.

File: batch.py [hunk 3]
```diff
@@ -163,54 +179,225 @@ def from_data_list(data_list):
...
+            "edge_index_substruct",
+            "x_substruct",
+            "overlap_context_substruct_idx",
+            "edge_attr_context",
+            "edge_index_context",
+            "x_context",
+        ]
+
+        for key in keys:
+            # print(key)
```
Comment (liketheflower) [inline batch.py:311]: This line can also be deleted.

**Review by liketheflower — COMMENTED** [review]
Looks good to me. Just remove several unused commented lines.
Another two updates:
1. Code format. I recently found one pretty good/easy to use tool to format all the python files at once. We can try to use it. The name is [black](https://pypi.org/project/black/). If you wanna use it:
- install:
```bash
pip install black
```
- Use:
```bash
jimmy@jimmys-MacBook-Pro contextSub % black contextSub 
reformatted /Users/jimmy/repos/jimmy/contextSub/contextSub/parse_result.py
reformatted /Users/jimmy/repos/jimmy/contextSub/contextSub/pretrain_edgepred.py
reformatted /Users/jimmy/repos/jimmy/contextSub/contextSub/pretrain_deepgraphinfomax.py
reformatted /Users/jimmy/repos/jimmy/contextSub/contextSub/pretrain_masking.py
reformatted /Users/jimmy/repos/jimmy/contextSub/contextSub/finetune_mutag_ptc.py
reformatted /Users/jimmy/repos/jimmy/contextSub/contextSub/pretrain_supervised.py
reformatted /Users/jimmy/repos/jimmy/contextSub/contextSub/dataloader.py
reformatted /Users/jimmy/repos/jimmy/contextSub/contextSub/finetune.py
reformatted /Users/jimmy/repos/jimmy/contextSub/contextSub/util.py
reformatted /Users/jimmy/repos/jimmy/contextSub/contextSub/splitters.py
All done! ✨ 🍰 ✨
10 files reformatted, 5 files left unchanged.
```
2. Code structure. I found all the code is in the same folder. I feel if we can put related files in one folder, it will be more clear. I can do this part after the code is merged. Since you mention that the code is runnable, please add the instruction about how to run the code in readme, so I can verify the code works after reorganize the code.
Thanks.

**Pull request merged**